Use the library name from the POM file of the dependency

rekire opened the pull request:
I want to have the clean name of the dependency if set in the POM file. My PR will use the name and will fallback when the name tag is not set in the POM file to the artifact group and artifact id.
Fixes #184

caller9 commented on oss-licenses-plugin/src/main/groovy/com/google/android/gms/oss/licenses/plugin/LicensesTask.groovy:
```diff
@@ -255,14 +259,22 @@ class LicensesTask extends DefaultTask {
-    protected void appendLicense(String key, byte[] content) {
-        if (licensesMap.containsKey(key)) {
+    protected void appendDependency(String dependency, byte[] license) {
+        String licenseText = new String(license)
```
Need to specify UTF_8 charset

rekire replied:
done

caller9 reviewed (changes requested):
This CL breaks unit tests in LicenseTaskTest. Please fix existing tests and add tests that verify the new functionality works as intended.

rekire commented:
I will fix the tests asap, tbh I forgot to run them

rekire commented on oss-licenses-plugin/src/main/groovy/com/google/android/gms/oss/licenses/plugin/LicensesTask.groovy:
```diff
@@ -216,16 +217,19 @@ class LicensesTask extends DefaultTask {
 
-        String licenseKey = "${group}:${name}"
+        String libraryName = rootNode.name
+        if (libraryName == null || libraryName == "") {
```
done

rekire commented:
I fixed the tests and your remarks. There was one broken test file (`groupB/bcd/deps2.pom`) is was a copy of `groupA/deps1.pom` instead of a different test case, I fixed that. Could you please review why I had to change the expected output in `testAddLicensesFromPom_withMultiple` to pass the test? I do not get why this was required.

Thank you for the review!

caller9 commented on oss-licenses-plugin/src/main/groovy/com/google/android/gms/oss/licenses/plugin/LicensesTask.groovy:
```diff
@@ -216,16 +217,19 @@ class LicensesTask extends DefaultTask {
         }
 
-        String licenseKey = "${group}:${name}"
+        String libraryName = rootNode.name
```
I did some further testing and unfortunately the name field is not sufficiently unique to provide a lookup. Maybe just switch to add the test fixes and deduplication here to #190. I'm OOO next week, but can take a look when I return.

In my testing with github.com/googlesamples/task-interop using this PR, two sets of SDKs collided on their name field which hides their licenses in an unpredictable way. You can see this by looking at app/build/generated/third_party_licenses/res/raw/third_party_licenses_metadata and comparing the output before and after this PR.

Example 1: https://maven.google.com/web/index.html?q=androidx.ac#androidx.activity:activity:1.3.1 and https://maven.google.com/web/index.html?q=savedstat#androidx.savedstate:savedstate:1.1.0 both have a POM Name of "Activity". This clobbers the licenses for one of them as they share the same key.

Example 2: https://repo1.maven.org/maven2/org/jetbrains/kotlinx/kotlinx-coroutines-core-metadata/1.4.2/kotlinx-coroutines-core-metadata-1.4.2.pom and https://repo1.maven.org/maven2/org/jetbrains/kotlinx/kotlinx-coroutines-android/1.5.1/kotlinx-coroutines-android-1.5.1.pom both have a POM Name of "kotlinx-coroutines-android". Again one will always clobber the other. 

There are probably a lot of examples of this sort of collision as the name isn't guaranteed to be unique while the artifact's maven coordinates are.

rekire replied:
That is interesting what you find out. I will review this and fix my new bugs.

rekire replied:
Ah crap. Resolving conversation does not submit the comment.
I did not intend to dedublicate the licenses, so I rolled back my changes a bit that the library's group and artifact id is used again. I merged #190 into this so everything should be fine now

rekire commented:
Changed the code and merged my other PR into this one. I'm looking forward new feedback

rekire commented on oss-licenses-plugin/src/test/java/com/google/android/gms/oss/licenses/plugin/LicensesTaskTest.java:
```diff
@@ -172,16 +172,14 @@ public void testAddLicensesFromPom_withMultiple() throws IOException {
     String content = new String(Files.readAllBytes(licensesTask.getLicenses().toPath()), UTF_8);
     String expected =
         "http://www.opensource.org/licenses/mit-license.php"
```
That was not intended in first place and caused the other bugs you found.

rekire commented on oss-licenses-plugin/src/test/java/com/google/android/gms/oss/licenses/plugin/LicensesTaskTest.java:
```diff
@@ -172,16 +172,14 @@ public void testAddLicensesFromPom_withMultiple() throws IOException {
-    assertTrue(licensesTask.licensesMap.containsKey("groupE:deps5 Apache License 2.0"));
+    assertTrue(licensesTask.licensesMap.containsKey("groupA deps1"));
+    assertTrue(licensesTask.licensesMap.containsKey("groupE deps5 MIT License"));
+    assertTrue(licensesTask.licensesMap.containsKey("groupE deps5 Apache License 2.0"));
```
I changed my code so that this is not required anymore

rekire commented:
@caller9 when I see this correctly then has my patch already landed in multiple Google Apps like the PlayStore, is there any timeframe when I can expect to see this in a public release?

rekire commented:
My patch went live in 0.10.5 on [February 28](https://developers.google.com/android/guides/releases#february_28_2022)
